contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,215 @@
+            revert Errors.InsufficientBuybackOutputAmount(feeToken, buybackToken, outputAmount, minOutputAmount);
+        }
+
+        _buybackFees(feeToken, buybackToken, batchSize, outputAmount, availableFeeAmount);
+    }
+
+    function getOutputAmount(
+        address[] calldata markets,
+        address feeToken,
+        address buybackToken
+    ) external view returns (uint256) {
+        uint256 batchSize = _getBatchSize(buybackToken);
+        _validateBuybackToken(batchSize, buybackToken);
+
+        uint256 feeAmount;
+        FeeAmounts memory feeAmounts;
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
         for (uint256 i; i < markets.length; i++) {
-            FeeUtils.claimFees(
-                dataStore,
-                eventEmitter,
-                markets[i],
-                tokens[i],
-                receiver
-            );
+            if (markets[i] == address(0)) {
+                feeAmount = vaultV1.feeReserves(feeToken);
+                feeAmounts = _getFeeAmounts(1, feeAmount);
+            } else {
+                feeAmount = _getUint(Keys.claimableFeeAmountKey(markets[i], feeToken));
+                feeAmounts = _getFeeAmounts(2, feeAmount);
+            }
+
+            feeAmount = buybackToken == gmx ? feeAmounts.gmx : feeAmounts.wnt;
+            availableFeeAmount = availableFeeAmount + feeAmount;
+        }
+
+        uint256 maxFeeTokenAmount = _getMaxFeeTokenAmount(feeToken, buybackToken, batchSize);
+        if (availableFeeAmount >= maxFeeTokenAmount) {
+            return maxFeeTokenAmount;
+        } else {
+            return availableFeeAmount;
+        }
+    }
+
+    function _buybackFees(
+        address feeToken,
+        address buybackToken,
+        uint256 batchSize,
+        uint256 buybackAmount,
+        uint256 availableFeeAmount
+    ) private {
+        IERC20(feeToken).safeTransfer(msg.sender, buybackAmount);
+        IERC20(buybackToken).safeTransferFrom(msg.sender, address(this), batchSize);
+
+        _setAvailableFeeAmount(feeToken, buybackToken, availableFeeAmount - buybackAmount);
+    }
+
+    function _incrementAvailableFeeAmounts(uint256 version, address feeToken, uint256 feeAmount) private {
+        address wnt = dataStore.getAddress(Keys.WNT);
+
+        FeeAmounts memory feeAmounts = _getFeeAmounts(version, feeAmount);
+
+        uint256 availableFeeAmountGmx = _getAvailableFeeAmount(feeToken, gmx) + feeAmounts.gmx;
+        uint256 availableFeeAmountWnt = _getAvailableFeeAmount(feeToken, wnt) + feeAmounts.wnt;
+
+        _setAvailableFeeAmount(feeToken, gmx, availableFeeAmountGmx);
+        _setAvailableFeeAmount(feeToken, wnt, availableFeeAmountWnt);
+    }
+
+    function _setAvailableFeeAmount(address feeToken, address buybackToken, uint256 availableFeeAmount) private {
+        dataStore.setUint(Keys.buybackAvailableFeeAmountKey(feeToken, buybackToken), availableFeeAmount);
```

Comment:
also wondering if it's necessary to create helpers that only call setUint/getUint, at first glance calling dataStore directly is clear/concise as well?

Reply:
i think this is fine, we do it in multiple other places as well: https://github.com/gmx-io/gmx-synthetics/blob/main/contracts/market/MarketUtils.sol#L1986

Reply:
maybe it's fine if these methods are later used in some other contracts, but in this case it only used inside this contract

it might be a matter of taste, i would personally prefer not to add more helpers, in this specific case they don't make code clearer in my view, but increase the code source size / adds some complexity to understand that code because i need to do an extra step to check what's inside `_setAvailableFeeAmount`

Reply:
discussed that this can be separately decided on